Replace CScriptID and CKeyID in CTxDestination with dedicated types

instagibbs opened the pull request:
The current usage seems to be an overloading of meanings. `CScriptID` is used in the wallet as a lookup key, as well as a destination, and `CKeyID` likewise. Instead, have all destinations be dedicated types.

New types:
`CScriptID`->`ScriptHash`
`CKeyID`->`PKHash`

instagibbs committed:
Replace CScriptID and CKeyID in CTxDestination with dedicated types

laanwj commented on src/script/standard.h:
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
+    using uint160::uint160;
+};
+
+struct SHash : public uint160
```
Concept ACK—I dislike the name `SHash` though, it sounds too general. I'd prefer a longer name that makes it more clear what is being hashed, maybe `ScriptHash`?

instagibbs replied:
done

sipa commented:
Concept ACK. I believe @ryanofsky suggested something like this before.

practicalswift commented on src/script/standard.h:
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
+{
+    ScriptHash() : uint160() {}
+    explicit ScriptHash(const uint160& hash) : uint160(hash) {}
+    explicit ScriptHash(const CScript& script);
```
Nit: Make sure parameter names match between declaration and definition :-)

instagibbs commented:
Not sure how to read appveyor error tbh

MarcoFalke commented:
The appveyor sync failure was fixed earlier this week

Sjors commented:
Concept ACK. Maybe enumerate the new types in the PR description?

Shouldn't `CKeyID GetID()` in `pubkey.h` also return a  `PKHash `? That breaks a few other things from the looks of it, but otherwise it seems inconsistent.

Sjors commented on src/script/standard.h:
```diff
@@ -73,6 +73,22 @@ class CNoDestination {
+{
+    PKHash() : uint160() {}
+    explicit PKHash(const uint160& hash) : uint160(hash) {}
+    explicit PKHash(const CPubKey& pubkey);
```
Inline definition might be more clear but I don't want to upset any strong c++ preferences out there.

instagibbs replied:
It's a mixed bag here(definition for other constructors are in .cpp), I'll leave as-is unless someone else chimes in

stevenroose commented:
utACK 3330398ab5df6f67859307339c0bb71c6b687220

instagibbs commented:
rebased

instagibbs commented:
rebased and added description of new destination types in OP @Sjors

ryanofsky reviewed (approved):
Conditional utACK 51482131c645ad565ad032a57bd105ede65f9b86 if fixes are added for travis failures. I'm guessing this will just require rebasing and adding some more CKeyID/CScriptID replacements.

This change was also suggested a while ago in: https://github.com/bitcoin/bitcoin/pull/11403#discussion_r155821355 and I think it could be simplified a little by taking the other suggestion in that comment and making ScriptHash inherit from CScriptID and PKHash inherit from CKeyID. This would avoid the need to add explicit conversions several places in this PR. I think it's important for conversions from CScriptID to ScriptHash and CKeyID to PKHash to be explicit, but it seems fine for conversions in the opposite direction to be implicit.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -3768,7 +3768,7 @@ void CWallet::GetKeyBirthTimes(interfaces::Chain::Lock& locked_chain, std::map<C
     for (const auto& entry : mapKeyMetadata) {
         if (entry.second.nCreateTime) {
-            mapKeyBirth[entry.first] = entry.second.nCreateTime;
+            mapKeyBirth[PKHash(entry.first)] = entry.second.nCreateTime;
```
done

instagibbs commented:
fixed a single dangling test case that was expecting a `PKHash`, and added a commit to change GetKeyBirthTimes to returning `CKeyID` instead of `CTxDestination`.

ryanofsky reviewed (approved):
utACK ddc21f2f2360a3d7e7acdddd4478822d47c44f7a. Changes since last review are fixing test case and switching GetKeyBirthTimes away from CTxDestination.

Sjors commented:
tACK ddc21f2. Compiled, opened a wallet and ran test suite and bench on macOS 10.14.4. Not sure what else to test here.

meshcollider commented on src/rpc/rawtransaction.cpp:
```diff
@@ -571,7 +571,7 @@ static UniValue decodescript(const JSONRPCRequest& request)
         // P2SH cannot be wrapped in a P2SH. If this script is already a P2SH,
         // don't return the address for a P2SH of the P2SH.
-        r.pushKV("p2sh", EncodeDestination(CScriptID(script)));
+        r.pushKV("p2sh", EncodeDestination(ScriptHash(CScriptID(script))));
```
`ScriptHash(script)`? Same on line 601

instagibbs replied:
fixed

meshcollider commented on src/test/script_standard_tests.cpp:
```diff
@@ -490,8 +490,8 @@ BOOST_AUTO_TEST_CASE(script_standard_IsMine)
 
-        CScript redeemscript = GetScriptForDestination(pubkeys[0].GetID());
-        CScript witnessscript = GetScriptForDestination(CScriptID(redeemscript));
+        CScript redeemscript = GetScriptForDestination(ScriptHash(PKHash(pubkeys[0])));
```
Why is there `ScriptHash` here?

instagibbs replied:
hmm seems I broke this test case, thanks

meshcollider commented on src/test/script_standard_tests.cpp:
```diff
@@ -598,7 +598,7 @@ BOOST_AUTO_TEST_CASE(script_standard_IsMine)
 
         CScript redeemScript = GetScriptForMultisig(2, {uncompressedPubkey, pubkeys[1]});
-        scriptPubKey = GetScriptForDestination(CScriptID(redeemScript));
+        scriptPubKey = GetScriptForDestination(ScriptHash(CScriptID(redeemScript)));
```
`ScriptHash(redeemScript)` again?

instagibbs replied:
fixed

Sjors replied:
A followup PR that would make this particular mistake fail the test would be nice though.

instagibbs committed:
GetKeyBirthTimes should return key ids, not destinations

instagibbs commented:
all comments addressed

ryanofsky reviewed (approved):
utACK 78e407ad0c26190a22de1bc8ed900164a44a36c3. Only changes are removing extra CScriptID()s and fixing the test case.

Sjors commented:
utACK 78e407a

Sjors reviewed (approved):
utACK 78e407a

meshcollider reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/15452/commits/78e407ad0c26190a22de1bc8ed900164a44a36c3

By removal of the `// TODO: include scripts in GetKeyBirthTimes() output instead of separate` comment, I assume you think that shouldn't be done, rather than because you implemented it.

instagibbs commented:
@meshcollider payoff was very low, and descriptor wallets would have changed whatever someone was going to do
